File: app/src/main/java/com/example/appestable/domain/PermissionPolicy.kt
package com.example.appestable.domain

import com.example.appestable.data.Persona
import com.example.appestable.data.RolViaje

object PermissionPolicy {

    fun canViewFamilia(session: SessionContext, familiaId: Int): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR -> true
        RolViaje.JEFE_FAMILIA, RolViaje.MIEMBRO -> session.familiaId == familiaId
    }

    fun canCreateViaje(session: SessionContext): Boolean =
        session.rol == RolViaje.ORGANIZADOR

    fun canAddPersona(session: SessionContext, familiaId: Int): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR -> true
        RolViaje.JEFE_FAMILIA -> session.familiaId == familiaId
        RolViaje.MIEMBRO -> false
    }

    fun canDeletePersona(session: SessionContext, persona: Persona): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR -> persona.rol != RolViaje.ORGANIZADOR
        RolViaje.JEFE_FAMILIA -> session.familiaId == persona.familiaId && persona.rol == RolViaje.MIEMBRO
        RolViaje.MIEMBRO -> false
    }

    fun canCreateActividad(session: SessionContext): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR, RolViaje.JEFE_FAMILIA -> true
        RolViaje.MIEMBRO -> false
    }

    fun canSelectParticipante(session: SessionContext, persona: Persona): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR -> true
        RolViaje.JEFE_FAMILIA -> session.familiaId == persona.familiaId
        RolViaje.MIEMBRO -> false
    }

    fun canEditParticipacion(session: SessionContext, familiaId: Int): Boolean = when (session.rol) {
        RolViaje.ORGANIZADOR -> true
        RolViaje.JEFE_FAMILIA -> session.familiaId == familiaId
        RolViaje.MIEMBRO -> false
    }

    fun canViewResumenGlobal(session: SessionContext): Boolean =
        session.rol == RolViaje.ORGANIZADOR
}

File: flutter_app/lib/domain/permission_policy.dart
import '../models/persona.dart';
import '../models/rol_viaje.dart';
import 'session_context.dart';

class PermissionPolicy {
  static bool canViewFamilia(SessionContext session, int familiaId) {
    switch (session.rol) {
      case RolViaje.organizador:
        return true;
      case RolViaje.jefeFamilia:
      case RolViaje.miembro:
        return session.familiaId == familiaId;
    }
  }

  static bool canCreateViaje(SessionContext session) =>
      session.rol == RolViaje.organizador;

  static bool canAddPersona(SessionContext session, int familiaId) {
    switch (session.rol) {
      case RolViaje.organizador:
        return true;
      case RolViaje.jefeFamilia:
        return session.familiaId == familiaId;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canDeletePersona(SessionContext session, Persona persona) {
    switch (session.rol) {
      case RolViaje.organizador:
        return persona.rol != RolViaje.organizador;
      case RolViaje.jefeFamilia:
        return session.familiaId == persona.familiaId &&
            persona.rol == RolViaje.miembro;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canCreateActividad(SessionContext session) {
    switch (session.rol) {
      case RolViaje.organizador:
      case RolViaje.jefeFamilia:
        return true;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canDeleteActividad(SessionContext session) {
    switch (session.rol) {
      case RolViaje.organizador:
      case RolViaje.jefeFamilia:
        return true;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canSelectParticipante(SessionContext session, Persona persona) {
    switch (session.rol) {
      case RolViaje.organizador:
        return true;
      case RolViaje.jefeFamilia:
        return session.familiaId == persona.familiaId;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canEditParticipacion(SessionContext session, int familiaId) {
    switch (session.rol) {
      case RolViaje.organizador:
        return true;
      case RolViaje.jefeFamilia:
        return session.familiaId == familiaId;
      case RolViaje.miembro:
        return false;
    }
  }

  static bool canViewResumenGlobal(SessionContext session) =>
      session.rol == RolViaje.organizador;
}